<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Getting data in and out of R</title>

<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="libs/navigation-1.1/tabsets.js"></script>
<link href="libs/highlightjs-1.1/default.css" rel="stylesheet" />
<script src="libs/highlightjs-1.1/highlight.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-68219208-1', 'auto');
  ga('send', 'pageview');

</script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="libs/local/main.css" type="text/css" />
<link rel="stylesheet" href="libs/local/nav.css" type="text/css" />
<link rel="stylesheet" href="//netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->





<header>
  <div class="nav">
    <a class="nav-logo" href="index.html">
      <img src="static/img/stat545-logo-s.png" width="70px" height="70px"/>
    </a>
    <ul>
      <li class="home"><a href="index.html">Home</a></li>
      <li class="faq"><a href="faq.html">FAQ</a></li>
      <li class="syllabus"><a href="syllabus.html">Syllabus</a></li>
      <li class="topics"><a href="topics.html">Topics</a></li>
      <li class="people"><a href="people.html">People</a></li>
    </ul>
  </div>
</header>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Getting data in and out of R</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#file-io-overview">File I/O overview</a></li>
<li><a href="#load-the-tidyverse-and-forcats">Load the tidyverse and forcats</a></li>
<li><a href="#locate-the-gapminder-data">Locate the Gapminder data</a></li>
<li><a href="#bring-rectangular-data-in">Bring rectangular data in</a></li>
<li><a href="#compute-something-worthy-of-export">Compute something worthy of export</a></li>
<li><a href="#write-rectangular-data-out">Write rectangular data out</a></li>
<li><a href="#invertibility">Invertibility</a></li>
<li><a href="#reordering-the-levels-of-the-country-factor">Reordering the levels of the country factor</a></li>
<li><a href="#saverds-and-readrds"><code>saveRDS()</code> and <code>readRDS()</code></a></li>
<li><a href="#retaining-factor-levels-upon-re-import">Retaining factor levels upon re-import</a></li>
<li><a href="#dput-and-dget"><code>dput()</code> and <code>dget()</code></a></li>
<li><a href="#other-types-of-objects-to-use-dput-or-saverds-on">Other types of objects to use <code>dput()</code> or <code>saveRDS()</code> on</a></li>
<li><a href="#clean-up">Clean up</a></li>
<li><a href="#pitfalls-of-delimited-files">Pitfalls of delimited files <a id="pitfalls"></a></a></li>
<li><a href="#references">References</a></li>
</ul>
</div>

<div id="file-io-overview" class="section level3">
<h3>File I/O overview</h3>
<p>We’ve been loading the Gapminder data as a data.frame from the <code>gapminder</code> data package. We haven’t been explicitly writing any data or derived results to file. In real life, you’ll bring rectangular data into and out of R all the time. Sometimes you’ll need to do same for non-rectangular objects.</p>
<p>How do you do this? What issues should you think about?</p>
<div id="data-import-mindset" class="section level4">
<h4>Data import mindset</h4>
<p>Data import generally feels one of two ways:</p>
<ul>
<li>“Surprise me!” This is the attitude you must adopt when you first get a dataset. You are just happy to import without an error. You start to explore. You discover flaws in the data and/or the import. You address them. Lather, rinse, repeat.</li>
<li>“Another day in paradise.” This is the attitude when you bring in a tidy dataset you have maniacally cleaned in one or more cleaning scripts. There should be no surprises. You should express your expectations about the data in formal assertions at the very start of these downstream scripts.</li>
</ul>
<p>In the second case, and as the first cases progresses, you actually know a lot about how the data is / should be. My main import advice: <strong>use the arguments of your import function to get as far as you can, as fast as possible</strong>. Novice code often has a great deal of unnecessary post import fussing around. Read the docs for the import functions and take maximum advantage of the arguments to control the import.</p>
</div>
<div id="data-export-mindset" class="section level4">
<h4>Data export mindset</h4>
<p>There will be many occasions when you need to write data from R. Two main examples:</p>
<ul>
<li>a tidy ready-to-analyze dataset that you heroically created from messy data</li>
<li>a numerical result from data aggregation or modelling or statistical inference</li>
</ul>
<p>First tip: <strong>today’s outputs are tomorrow’s inputs</strong>. Think back on all the pain you have suffered importing data and don’t inflict such pain on yourself!</p>
<p>Second tip: don’t be too cute or clever. A plain text file that is readable by a human being in a text editor should be your default until you have <strong>actual proof</strong> that this will not work. Reading and writing to exotic or proprietary formats will be the first thing to break in the future or on a different computer. It also creates barriers for anyone who has a different toolkit than you do. Be software-agnostic. Aim for future-proof and moron-proof.</p>
<p>How does this fit with our emphasis on dynamic reporting via R Markdown? There is a time and place for everything. There are projects and documents where the scope and personnel will allow you to geek out with <code>knitr</code> and R Markdown. But there are lots of good reasons why (parts of) an analysis should not (only) be embedded in a dynamic report. Maybe you are just doing data cleaning to produce a valid input dataset. Maybe you are making a small but crucial contribution to a giant multi-author paper. Etc. Also remember there are other tools and workflows for making something reproducible. I’m looking at you, <a href="http://kbroman.github.io/minimal_make/">make</a>.</p>
</div>
</div>
<div id="load-the-tidyverse-and-forcats" class="section level3">
<h3>Load the tidyverse and forcats</h3>
<p>The main function we will be using is readr, which provides drop-in substitutes for <code>read.table()</code> and friends. However, to make some points about data export and import, it is nice to reorder factor levels. For that, we will also load and use the forcats package.</p>
<pre class="r"><code>library(tidyverse)
## + ggplot2 2.2.1             Date: 2017-10-29
## + tibble  1.3.4                R: 3.4.1
## + tidyr   0.7.1               OS: macOS Sierra 10.12.6
## + readr   1.1.1              GUI: X11
## + purrr   0.2.3.9000      Locale: en_CA.UTF-8
## + dplyr   0.7.4               TZ: America/Vancouver
## + stringr 1.2.0.9000      
## + forcats 0.2.0
## Warning: package &#39;dplyr&#39; was built under R version 3.4.2
## ── Conflicts ────────────────────────────────────────────────────
## * filter(),  from dplyr, masks stats::filter()
## * lag(),     from dplyr, masks stats::lag()
library(forcats)</code></pre>
</div>
<div id="locate-the-gapminder-data" class="section level3">
<h3>Locate the Gapminder data</h3>
<p>We could load the data from the package as usual, but instead we will load it from tab delimited file. The gapminder package includes the data normally found in the <code>gapminder</code> data frame as a <code>.tsv</code>. So let’s get the path to that file on <em>your</em> system.</p>
<pre class="r"><code>(gap_tsv &lt;- system.file(&quot;gapminder.tsv&quot;, package = &quot;gapminder&quot;))
## [1] &quot;/Users/jenny/resources/R/library/gapminder/gapminder.tsv&quot;</code></pre>
</div>
<div id="bring-rectangular-data-in" class="section level3">
<h3>Bring rectangular data in</h3>
<p>The workhorse data import function of readr is <code>read_delim()</code>. Here we’ll use a variant, <code>read_tsv()</code>, that anticipates tab-delimited data:</p>
<pre class="r"><code>gapminder &lt;- read_tsv(gap_tsv)
## Parsed with column specification:
## cols(
##   country = col_character(),
##   continent = col_character(),
##   year = col_integer(),
##   lifeExp = col_double(),
##   pop = col_integer(),
##   gdpPercap = col_double()
## )
str(gapminder, give.attr = FALSE)
## Classes &#39;tbl_df&#39;, &#39;tbl&#39; and &#39;data.frame&#39;:    1704 obs. of  6 variables:
##  $ country  : chr  &quot;Afghanistan&quot; &quot;Afghanistan&quot; &quot;Afghanistan&quot; &quot;Afghanistan&quot; ...
##  $ continent: chr  &quot;Asia&quot; &quot;Asia&quot; &quot;Asia&quot; &quot;Asia&quot; ...
##  $ year     : int  1952 1957 1962 1967 1972 1977 1982 1987 1992 1997 ...
##  $ lifeExp  : num  28.8 30.3 32 34 36.1 ...
##  $ pop      : int  8425333 9240934 10267083 11537966 13079460 14880372 12881816 13867957 16317921 22227415 ...
##  $ gdpPercap: num  779 821 853 836 740 ...</code></pre>
<p>For full flexibility re: specifying the delimiter, you can always use <code>readr::read_delim()</code>.</p>
<p>There’s a similar convenience wrapper for comma-separated values, <code>read_csv()</code>.</p>
<p>The most noticeable difference between the readr functions and base is that readr does NOT convert strings to factors by default. In the grand scheme of things, this is better default behavior, although we go ahead and convert them to factor here. Do not be deceived – in general, you will do less post-import fussing if you use readr.</p>
<pre class="r"><code>gapminder &lt;- gapminder %&gt;%
  mutate(country = factor(country),
         continent = factor(continent))
str(gapminder)
## Classes &#39;tbl_df&#39;, &#39;tbl&#39; and &#39;data.frame&#39;:    1704 obs. of  6 variables:
##  $ country  : Factor w/ 142 levels &quot;Afghanistan&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ continent: Factor w/ 5 levels &quot;Africa&quot;,&quot;Americas&quot;,..: 3 3 3 3 3 3 3 3 3 3 ...
##  $ year     : int  1952 1957 1962 1967 1972 1977 1982 1987 1992 1997 ...
##  $ lifeExp  : num  28.8 30.3 32 34 36.1 ...
##  $ pop      : int  8425333 9240934 10267083 11537966 13079460 14880372 12881816 13867957 16317921 22227415 ...
##  $ gdpPercap: num  779 821 853 836 740 ...</code></pre>
<div id="bring-rectangular-data-in-summary" class="section level4">
<h4>Bring rectangular data in – summary</h4>
<p>Default to <code>readr::read_delim()</code> and friends. Use the arguments!</p>
<p>The Gapminder data is too clean and simple to show off the great features of readr, so I encourage you to check out the vignette on <a href="https://cran.r-project.org/web/packages/readr/vignettes/column-types.html">column types</a>. There are many variable types that you will be able to parse correctly upon import, thereby eliminating a great deal of post-import fussing.</p>
</div>
</div>
<div id="compute-something-worthy-of-export" class="section level3">
<h3>Compute something worthy of export</h3>
<p>We need compute something worth writing to file. Let’s create a country-level summary of maximum life expectancy.</p>
<pre class="r"><code>gap_life_exp &lt;- gapminder %&gt;%
  group_by(country, continent) %&gt;% 
  summarise(life_exp = max(lifeExp)) %&gt;% 
  ungroup()
gap_life_exp
## # A tibble: 142 x 3
##        country continent life_exp
##         &lt;fctr&gt;    &lt;fctr&gt;    &lt;dbl&gt;
##  1 Afghanistan      Asia   43.828
##  2     Albania    Europe   76.423
##  3     Algeria    Africa   72.301
##  4      Angola    Africa   42.731
##  5   Argentina  Americas   75.320
##  6   Australia   Oceania   81.235
##  7     Austria    Europe   79.829
##  8     Bahrain      Asia   75.635
##  9  Bangladesh      Asia   64.062
## 10     Belgium    Europe   79.441
## # ... with 132 more rows</code></pre>
<p>The <code>gap_life_exp</code> data frame is an example of an intermediate result that we want to store for the future and for downstream analyses or visualizations.</p>
</div>
<div id="write-rectangular-data-out" class="section level3">
<h3>Write rectangular data out</h3>
<p>The workhorse export function for rectangular data in readr is <code>write_delim()</code> and friends. Let’s use <code>write_csv()</code> to get a comma-delimited file.</p>
<pre class="r"><code>write_csv(gap_life_exp, &quot;gap_life_exp.csv&quot;)</code></pre>
<p>Let’s look at the first few lines of <code>gap_life_exp.csv</code>. If you’re following along, you should be able to open this file or, in a shell, use <code>head</code> on it.</p>
<pre><code>country,continent,life_exp
Afghanistan,Asia,43.828
Albania,Europe,76.423
Algeria,Africa,72.301
Angola,Africa,42.731
Argentina,Americas,75.32</code></pre>
<p>This is pretty decent looking, though there is no visible alignment or separation into columns. Had we used the base function <code>read.csv()</code>, we would be seeing rownames and lots of quotes, unless we had explicitly shut that down. Nicer default behavior is the main reason we are using <code>readr::write_csv()</code> over <code>write.csv()</code>.</p>
<ul>
<li>It’s not really fair to complain about the lack of visible alignment. Remember we are <a href="https://twitter.com/vsbuffalo/statuses/358699162679787521">“writing data for computers”</a>. If you really want to browse around the file, use <code>View()</code> in RStudio or open it in Microsoft Excel (!) but don’t succumb to the temptation to start doing artisanal data manipulations there … get back to R and construct commands that you can re-run the next 15 times you import/clean/aggregate/export the same dataset. Trust me, it will happen.</li>
</ul>
</div>
<div id="invertibility" class="section level3">
<h3>Invertibility</h3>
<p>It turns out these self-imposed rules are often in conflict with one another</p>
<ul>
<li>Write to plain text files</li>
<li>Break analysis into pieces: the output of script <code>i</code> is an input for script <code>i + 1</code></li>
<li>Be the boss of factors: order the levels in a meaningful, usually non-alphabetical way</li>
<li>Avoid duplication of code and data</li>
</ul>
<p>Example: after performing the country-level summarization, we reorder the levels of the country factor, based on life expectancy. This reordering operation is conceptually important and must be embodied in R commands stored in a script. However, as soon as we write <code>gap_life_exp</code> to a plain text file, that meta-information about the countries is lost. Upon re-import with <code>read_delim()</code> and friends, we are back to alphabetically ordered factor levels. Any measure we take to avoid this loss immediately breaks another one of our rules.</p>
<p>So what do I do? I must admit I save (and re-load) R-specific binary files. Right after I save the plain text file. <a href="http://www.wisegeek.com/what-does-it-mean-to-wear-belt-and-suspenders.htm">Belt and suspenders</a>.</p>
<p>I have toyed with the idea of writing import helper functions for a specific project, that would re-order factor levels in principled ways. They could be defined in one file and called from many. This would also have a very natural implementation within <a href="http://carlboettiger.info/2012/05/06/research-workflow.html">a workflow where each analytical project is an R package</a>. But so far it has seemed too much like <a href="http://sethgodin.typepad.com/seths_blog/2005/03/dont_shave_that.html">yak shaving</a>. I’m intrigued by a recent discussion of putting such information in YAML frontmatter (see Martin Fenner blog post <a href="http://blog.datacite.org/using-yaml-frontmatter-with-csv/">Using YAML frontmatter with CSV</a>).</p>
</div>
<div id="reordering-the-levels-of-the-country-factor" class="section level3">
<h3>Reordering the levels of the country factor</h3>
<p>The topic of <a href="block029_factors.html">factor level reordering is covered elsewhere</a>, so let’s Just. Do. It. I reorder the country factor levels according to the life expectancy summary we’ve already computed.</p>
<pre class="r"><code>head(levels(gap_life_exp$country)) # alphabetical order
## [1] &quot;Afghanistan&quot; &quot;Albania&quot;     &quot;Algeria&quot;     &quot;Angola&quot;      &quot;Argentina&quot;  
## [6] &quot;Australia&quot;
gap_life_exp &lt;- gap_life_exp %&gt;% 
  mutate(country = fct_reorder(country, life_exp))
head(levels(gap_life_exp$country)) # in increasing order of maximum life expectancy
## [1] &quot;Sierra Leone&quot; &quot;Angola&quot;       &quot;Afghanistan&quot;  &quot;Liberia&quot;     
## [5] &quot;Rwanda&quot;       &quot;Mozambique&quot;
head(gap_life_exp)
## # A tibble: 6 x 3
##       country continent life_exp
##        &lt;fctr&gt;    &lt;fctr&gt;    &lt;dbl&gt;
## 1 Afghanistan      Asia   43.828
## 2     Albania    Europe   76.423
## 3     Algeria    Africa   72.301
## 4      Angola    Africa   42.731
## 5   Argentina  Americas   75.320
## 6   Australia   Oceania   81.235</code></pre>
<p>Note that the <strong>row order of <code>gap_life_exp</code> has not changed</strong>. I could choose to reorder the rows of the data frame if, for example, I was about to prepare a table to present to people. But I’m not, so I won’t.</p>
</div>
<div id="saverds-and-readrds" class="section level3">
<h3><code>saveRDS()</code> and <code>readRDS()</code></h3>
<p>If you have a data frame AND you have exerted yourself to rationalize the factor levels, you have my blessing to save it to file in a way that will preserve this hard work upon re-import. Use <code>saveRDS()</code>.</p>
<pre class="r"><code>saveRDS(gap_life_exp, &quot;gap_life_exp.rds&quot;)</code></pre>
<p><code>saveRDS()</code> serializes an R object to a binary file. It’s not a file you will able to open in an editor, diff nicely with Git(Hub), or share with non-R friends. It’s a special purpose, limited use function that I use in specific situations.</p>
<p>The opposite of <code>saveRDS()</code> is <code>readRDS()</code>. You must assign the return value to an object. I highly recommend you assign back to the same name as before. Why confuse yourself?!?</p>
<pre class="r"><code>rm(gap_life_exp)
gap_life_exp
## Error in eval(expr, envir, enclos): object &#39;gap_life_exp&#39; not found
gap_life_exp &lt;- readRDS(&quot;gap_life_exp.rds&quot;)
gap_life_exp
## # A tibble: 142 x 3
##        country continent life_exp
##         &lt;fctr&gt;    &lt;fctr&gt;    &lt;dbl&gt;
##  1 Afghanistan      Asia   43.828
##  2     Albania    Europe   76.423
##  3     Algeria    Africa   72.301
##  4      Angola    Africa   42.731
##  5   Argentina  Americas   75.320
##  6   Australia   Oceania   81.235
##  7     Austria    Europe   79.829
##  8     Bahrain      Asia   75.635
##  9  Bangladesh      Asia   64.062
## 10     Belgium    Europe   79.441
## # ... with 132 more rows</code></pre>
<p><code>saveRDS()</code> has more arguments, in particular <code>compress</code> for controlling compression, so read the help for more advanced usage. It is also very handy for saving non-rectangular objects, like a fitted regression model, that took a nontrivial amount of time to compute.</p>
<p>You will eventually hear about <code>save()</code> + <code>load()</code> and even <code>save.image()</code>. You may even see them in documentation and tutorials, but don’t be tempted. Just say no. These functions encourage unsafe practices, like storing multiple objects together and even entire workspaces. There are legitimate uses of these functions, but not in your typical data analysis.</p>
</div>
<div id="retaining-factor-levels-upon-re-import" class="section level3">
<h3>Retaining factor levels upon re-import</h3>
<p>Concrete demonstration of how non-alphabetical factor level order is lost with <code>write_delim()</code> / <code>read_delim()</code> workflows but maintained with <code>saveRDS()</code> / <code>readRDS()</code>.</p>
<pre class="r"><code>(country_levels &lt;- tibble(original = head(levels(gap_life_exp$country))))
## # A tibble: 6 x 1
##       original
##          &lt;chr&gt;
## 1 Sierra Leone
## 2       Angola
## 3  Afghanistan
## 4      Liberia
## 5       Rwanda
## 6   Mozambique
write_csv(gap_life_exp, &quot;gap_life_exp.csv&quot;)
saveRDS(gap_life_exp, &quot;gap_life_exp.rds&quot;)
rm(gap_life_exp)
head(gap_life_exp) # will cause error! proving gap_life_exp is really gone 
## Error in head(gap_life_exp): object &#39;gap_life_exp&#39; not found
gap_via_csv &lt;- read_csv(&quot;gap_life_exp.csv&quot;) %&gt;% 
  mutate(country = factor(country))
## Parsed with column specification:
## cols(
##   country = col_character(),
##   continent = col_character(),
##   life_exp = col_double()
## )
gap_via_rds &lt;- readRDS(&quot;gap_life_exp.rds&quot;)
country_levels &lt;- country_levels %&gt;% 
  mutate(via_csv = head(levels(gap_via_csv$country)),
         via_rds = head(levels(gap_via_rds$country)))
country_levels
## # A tibble: 6 x 3
##       original     via_csv      via_rds
##          &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;
## 1 Sierra Leone Afghanistan Sierra Leone
## 2       Angola     Albania       Angola
## 3  Afghanistan     Algeria  Afghanistan
## 4      Liberia      Angola      Liberia
## 5       Rwanda   Argentina       Rwanda
## 6   Mozambique   Australia   Mozambique</code></pre>
<p>Note how the original, post-reordering country factor levels are restored using the <code>saveRDS()</code> / <code>readRDS()</code> strategy but revert to alphabetical ordering using <code>write_csv()</code> / <code>read_csv()</code>.</p>
</div>
<div id="dput-and-dget" class="section level3">
<h3><code>dput()</code> and <code>dget()</code></h3>
<p>One last method of saving and restoring data deserves a mention: <code>dput()</code> and <code>dget()</code>. <code>dput()</code> offers this odd combination of features: it creates a plain text representation of an R object which still manages to be quite opaque. If you use the <code>file =</code> argument, <code>dput()</code> can write this representation to file but you won’t be tempted to actually read that thing. <code>dput()</code> creates an R-specific-but-not-binary representation. Let’s try it out.</p>
<pre class="r"><code>## first restore gap_life_exp with our desired country factor level order
gap_life_exp &lt;- readRDS(&quot;gap_life_exp.rds&quot;)
dput(gap_life_exp, &quot;gap_life_exp-dput.txt&quot;)</code></pre>
<p>Now let’s look at the first few lines of the file <code>gap_life_exp-dput.txt</code>.</p>
<pre><code>structure(list(country = structure(c(3L, 107L, 74L, 2L, 98L, 
138L, 128L, 102L, 49L, 125L, 26L, 56L, 96L, 47L, 75L, 85L, 18L, 
12L, 37L, 24L, 133L, 13L, 16L, 117L, 84L, 82L, 53L, 9L, 28L, 
120L, 22L, 104L, 114L, 109L, 115L, 23L, 73L, 97L, 66L, 71L, 15L, 
29L, 20L, 122L, 134L, 40L, 35L, 123L, 38L, 126L, 60L, 25L, 7L, 
39L, 59L, 141L, 86L, 140L, 51L, 63L, 64L, 52L, 121L, 135L, 132L, </code></pre>
<p>Huh? Don’t worry about it. Remember we are <a href="https://twitter.com/vsbuffalo/statuses/358699162679787521">“writing data for computers”</a>. The partner function <code>dget()</code> reads this representation back in.</p>
<pre class="r"><code>gap_life_exp_dget &lt;- dget(&quot;gap_life_exp-dput.txt&quot;)
country_levels &lt;- country_levels %&gt;% 
  mutate(via_dput = head(levels(gap_life_exp_dget$country)))
country_levels
## # A tibble: 6 x 4
##       original     via_csv      via_rds     via_dput
##          &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;        &lt;chr&gt;
## 1 Sierra Leone Afghanistan Sierra Leone Sierra Leone
## 2       Angola     Albania       Angola       Angola
## 3  Afghanistan     Algeria  Afghanistan  Afghanistan
## 4      Liberia      Angola      Liberia      Liberia
## 5       Rwanda   Argentina       Rwanda       Rwanda
## 6   Mozambique   Australia   Mozambique   Mozambique</code></pre>
<p>Note how the original, post-reordering country factor levels are restored using the <code>dput()</code> / <code>dget()</code> strategy.</p>
<p>But why on earth would you ever do this?</p>
<p>The main application of this is <a href="http://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example">the creation of highly portable, self-contained minimal examples</a>. For example, if you want to pose a question on a forum or directly to an expert, it might be required or just plain courteous to NOT attach any data files. You will need a monolithic, plain text blob that defines any necessary objects and has the necessary code. <code>dput()</code> can be helpful for producing the piece of code that defines the object. If you <code>dput()</code> without specifying a file, you can copy the return value from Console and paste into a script. Or you can write to file and copy from there or add R commands below.</p>
</div>
<div id="other-types-of-objects-to-use-dput-or-saverds-on" class="section level3">
<h3>Other types of objects to use <code>dput()</code> or <code>saveRDS()</code> on</h3>
<p>My special dispensation to abandon human-readable, plain text files is even broader than I’ve let on. Above, I give my blessing to store data.frames via <code>dput()</code> and/or <code>saveRDS()</code>, when you’ve done some rational factor level re-ordering. The same advice and mechanics apply a bit more broadly: you’re also allowed to use R-specific file formats to save vital non-rectangular objects, such as a fitted nonlinear mixed effects model or a classification and regression tree.</p>
</div>
<div id="clean-up" class="section level3">
<h3>Clean up</h3>
<p>We’ve written several files in this tutorial. Some of them are not of lasting value or have confusing filenames. I choose to delete them, while demonstrating some of the many functions R offers for interacting with the filesystem. It’s up to you whether you want to submit this command or not.</p>
<pre class="r"><code>file.remove(list.files(pattern = &quot;^gap_life_exp&quot;))
## [1] TRUE TRUE TRUE</code></pre>
</div>
<div id="pitfalls-of-delimited-files" class="section level3">
<h3>Pitfalls of delimited files <a id="pitfalls"></a></h3>
<p>If a delimited file contains fields where a human being has typed, be crazy paranoid because people do really nutty things. Especially people who aren’t in the business of programming and have never had to compute on text. Claim: a person’s regular expression skill is inversely proportional to the skill required to handle the files they create. Implication: if someone has never heard of regular expressions, prepare for lots of pain working with their files.</p>
<p>When the header fields (often, but not always, the variable names) or actual data contain the delimiter, it can lead to parsing and import failures. Two popular delimiters are the comma <code>,</code> and the TAB <code>\t</code> and humans tend to use these when typing. If you can design this problem away during data capture, such as by using a drop down menu on an input form, by all means do so. Sometimes this is impossible or undesirable and you must deal with fairly free form text. That’s a good time to allow/force text to be protected with quotes, because it will make parsing the delimited file go more smoothly.</p>
<p>Sometimes, instead of rigid tab-delimiting, whitespace is used as the delimiter. That is, in fact, the default for both <code>read.table()</code> and <code>write.table()</code>. Assuming you will write/read variable names from the first line (a.k.a. the <code>header</code> in <code>write.table()</code> and <code>read.table()</code>), they must be valid R variable names … or they will be coerced into something valid. So, for these two reasons, it is good practice to use “one word” variable names whenever possible. If you need to evoke multiple words, use <code>snake_case</code> or <code>camelCase</code> to cope. Example: the header entry for the field holding the subject’s last name should be <code>last_name</code> or <code>lastName</code> NOT <code>last name</code>. With the <code>readr</code> package, “column names are left as is, not munged into valid R identifiers (i.e. there is no <code>check.names = TRUE</code>)”. So you can get away with whitespace in variable names and yet I recommend that you do not.</p>
</div>
<div id="references" class="section level3">
<h3>References</h3>
<p><a href="http://r4ds.had.co.nz/data-import.html">Data import</a> chapter of <a href="http://r4ds.had.co.nz">R for Data Science</a> by Hadley Wickahm and Garrett Grolemund.</p>
<p>Nine simple ways to make it easier to (re)use your data by Ethan P White, Elita Baldridge, Zachary T. Brym, Kenneth J. Locey, Daniel J. McGlinn, Sarah R. Supp.</p>
<ul>
<li>First appeared here: PeerJ PrePrints 1:e7v2 <a href="http://dx.doi.org/10.7287/peerj.preprints.7v2" class="uri">http://dx.doi.org/10.7287/peerj.preprints.7v2</a></li>
<li>Published here: Ideas in Ecology and Evolution 6(2): 1?10, 2013. <a href="doi:10.4033/iee.2013.6b.6.f" class="uri">doi:10.4033/iee.2013.6b.6.f</a> <a href="http://library.queensu.ca/ojs/index.php/IEE/article/view/4608" class="uri">http://library.queensu.ca/ojs/index.php/IEE/article/view/4608</a></li>
<li>Section 4 “Use Standard Data Formats” is especially good reading.</li>
</ul>
<p>Tidy data by Hadley Wickham.</p>
<ul>
<li>In the Journal of Statistical Software Vol 59 (2014), Issue 10, 10.18637/jss.v059.i10: <a href="http://www.jstatsoft.org/article/view/v059i10" class="uri">http://www.jstatsoft.org/article/view/v059i10</a></li>
<li>PDF also available here: <a href="http://vita.had.co.nz/papers/tidy-data.pdf" class="uri">http://vita.had.co.nz/papers/tidy-data.pdf</a></li>
</ul>
<p>Data Manipulation with R <a href="http://ezproxy.library.ubc.ca/login?url=http://link.springer.com.ezproxy.library.ubc.ca/book/10.1007/978-0-387-74731-6/page/1">available via SpringerLink</a> by Phil Spector, Springer (2008) | <a href="http://www.stat.berkeley.edu/%7Espector/">author webpage</a> | <a href="http://books.google.com/books?id=grfuq1twFe4C&amp;lpg=PP1&amp;dq=data%2520manipulation%2520spector&amp;pg=PP1#v=onepage&amp;q=&amp;f=false">GoogleBooks search</a></p>
<ul>
<li>See Chapter 2 (“Reading and Writing Data”)</li>
</ul>
</div>

<div class="footer">
This work is licensed under the  <a href="http://creativecommons.org/licenses/by-nc/3.0/">CC BY-NC 3.0 Creative Commons License</a>.
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
